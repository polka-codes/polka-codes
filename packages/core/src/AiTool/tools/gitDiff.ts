// generated by polka.codes
import { z } from 'zod'
import { type FullToolInfoV2, PermissionLevel, type ToolHandler, type ToolInfoV2, ToolResponseType } from '../../tool'
import type { CommandProvider } from '../../tools'

export const toolInfo = {
  name: 'git_diff',
  description:
    'Get the git diff for the current repository. Can be used to get staged changes, unstaged changes, or changes between commits. By default, it returns unstaged changes.',
  parameters: z.object({
    staged: z
      .preprocess((val) => {
        if (typeof val === 'string') {
          const lower = val.toLowerCase()
          if (lower === 'false') return false
          if (lower === 'true') return true
        }
        return val
      }, z.boolean().optional().default(false))
      .describe('Get staged changes instead of unstaged changes.'),
    commitRange: z.string().optional().describe('The commit range to get the diff for (e.g., "main...HEAD").'),
    file: z.string().optional().describe('Get the diff for a specific file.'),
  }),
  permissionLevel: PermissionLevel.Read,
} as const satisfies ToolInfoV2

export const handler: ToolHandler<typeof toolInfo, CommandProvider> = async (provider, args) => {
  if (!provider.executeCommand) {
    return {
      type: ToolResponseType.Error,
      message: 'Not possible to execute command. Abort.',
    }
  }

  const { staged, file, commitRange } = toolInfo.parameters.parse(args)

  const commandParts = ['git', 'diff', '--no-color']
  if (staged) {
    commandParts.push('--staged')
  }
  if (commitRange) {
    commandParts.push(commitRange)
  }
  if (file) {
    commandParts.push('--', file)
  }

  const command = commandParts.join(' ')
  try {
    const result = await provider.executeCommand(command, false)
    if (result.exitCode === 0) {
      if (!result.stdout.trim()) {
        return {
          type: ToolResponseType.Reply,
          message: 'No diff found.',
        }
      }
      return {
        type: ToolResponseType.Reply,
        message: `<diff file="${file ?? 'all'}">\n${result.stdout}\n</diff>`,
      }
    }
    return {
      type: ToolResponseType.Error,
      message: `\`${command}\` exited with code ${result.exitCode}:\n${result.stderr}`,
    }
  } catch (error) {
    return {
      type: ToolResponseType.Error,
      message: error instanceof Error ? error.message : String(error),
    }
  }
}

export const isAvailable = (provider: CommandProvider): boolean => {
  return !!provider.executeCommand
}

export default {
  ...toolInfo,
  handler,
  isAvailable,
} satisfies FullToolInfoV2
