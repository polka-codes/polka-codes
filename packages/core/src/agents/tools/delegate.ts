import { z } from 'zod'

import { type AgentInfo, agentTool } from '../types'

export const delegateTool = (agents: AgentInfo[]) => {
  const agentInputSchemas = agents.map((agent) =>
    z.object({
      agentName: z.literal(agent.name),
      task: z.string().describe('The task to be completed by the target agent'),
      ...(agent.contextSchema ? { context: agent.contextSchema.describe('The context information for the task') } : {}),
    }),
  )

  let inputSchema: z.ZodSchema
  if (agents.length === 0) {
    throw new Error('At least one agent is required to delegate a task')
  }
  if (agents.length === 1) {
    inputSchema = agentInputSchemas[0]
  } else {
    inputSchema = z.union(agentInputSchemas as any)
  }

  return agentTool({
    id: 'delegate',
    description: 'Delegate a task to another agent to execute and receive the result back',
    inputSchema: inputSchema.describe('The input for the task'),
    outputSchema: z.object({
      status: z.enum(['success', 'failure']).describe('The status of the task'),
      output: z.string().describe('The output of the task'),
    }),
    execute: async (input, polka) => {
      try {
        const result = await polka.startTask(input.agentName, input.task, input.context)
        if (typeof result === 'string') {
          return {
            status: 'success' as const,
            output: result,
          }
        }
        return {
          status: 'failure' as const,
          output: result.error,
        }
      } catch (error: any) {
        return {
          status: 'failure' as const,
          output: 'message' in error ? error.message : (error?.toString() ?? JSON.stringify(error)),
        }
      }
    },
  })
}
