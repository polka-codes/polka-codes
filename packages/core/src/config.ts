import { z } from 'zod'

const providerModelSchema = z.object({
  provider: z.string().optional(),
  model: z.string().optional(),
  parameters: z.record(z.string(), z.any()).optional(),
})

const agentSchema = providerModelSchema.extend({
  initialContext: z
    .object({
      maxFileCount: z.number().int().positive().optional(),
      excludes: z.array(z.string()).optional(),
    })
    .optional(),
})

export const configSchema = z
  .object({
    agent: z.string().optional(),
    prices: z
      .record(
        z.string(), // provider
        z.record(
          z.string(), // model
          z.object({
            inputPrice: z.number().optional(),
            outputPrice: z.number().optional(),
            cacheWritesPrice: z.number().optional(),
            cacheReadsPrice: z.number().optional(),
          }),
        ),
      )
      .optional(),
    providers: z
      .record(
        z.string(),
        z.object({
          apiKey: z.string().optional(),
          defaultModel: z.string().optional(),
          defaultParameters: z.record(z.string(), z.any()).optional(),
        }),
      )
      .optional(),
    defaultProvider: z.string().optional(),
    defaultModel: z.string().optional(),
    defaultParameters: z.record(z.string(), z.any()).optional(),
    maxMessageCount: z.number().int().positive().optional(),
    budget: z.number().positive().optional(),
    scripts: z
      .record(
        z.string(),
        z.string().or(
          z.object({
            command: z.string(),
            description: z.string(),
          }),
        ),
      )
      .optional(),
    agents: z.record(z.string(), agentSchema).optional(),
    commands: z
      .object({
        default: providerModelSchema.optional(),
      })
      .catchall(providerModelSchema)
      .optional(),
    rules: z.array(z.string()).optional().or(z.string()).optional(),
    excludeFiles: z.array(z.string()).optional(),
  })
  .strict()

export type Config = z.infer<typeof configSchema>
