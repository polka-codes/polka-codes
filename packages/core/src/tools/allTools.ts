export { default as askFollowupQuestion } from './askFollowupQuestion'
export { default as attemptCompletion } from './attemptCompletion'
export { default as executeCommand } from './executeCommand'
export { default as listCodeDefinitionNames } from './listCodeDefinitionNames'
export { default as listFiles } from './listFiles'
export { default as readFile } from './readFile'
export { default as replaceInFile } from './replaceInFile'
export { default as searchFiles } from './searchFiles'
export { default as writeToFile } from './writeToFile'
export { default as handOver } from './handOver'
export { default as removeFile } from './removeFile'
export { default as renameFile } from './renameFile'
