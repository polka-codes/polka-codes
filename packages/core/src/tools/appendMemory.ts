import { z } from 'zod'
import { type FullToolInfo, type ToolHandler, type ToolInfo, ToolResponseType } from '../tool'
import type { MemoryProvider } from './provider'

export const toolInfo = {
  name: 'appendMemory',
  description: 'Appends content to a memory topic.',
  parameters: z.object({
    topic: z.string().optional().describe('The topic to append to in memory. Defaults to ":default:".'),
    content: z.string().describe('The content to append.'),
  }),
} as const satisfies ToolInfo

export const handler: ToolHandler<typeof toolInfo, MemoryProvider> = async (provider, args) => {
  const { topic, content } = toolInfo.parameters.parse(args)
  await provider.append(topic || ':default:', content)
  return {
    type: ToolResponseType.Reply,
    message: {
      type: 'text',
      value: `Content appended to memory topic '${topic || ':default:'}'.`,
    },
  }
}

export default {
  ...toolInfo,
  handler,
} satisfies FullToolInfo
