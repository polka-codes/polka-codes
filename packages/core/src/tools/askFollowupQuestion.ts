import { z } from 'zod'
import { type FullToolInfoV2, PermissionLevel, type ToolHandler, type ToolInfoV2, ToolResponseType } from '../tool'
import type { InteractionProvider } from './provider'

const questionObject = z.object({
  prompt: z.string().describe('The text of the question.').meta({ usageValue: 'question text here' }),
  options: z
    .array(z.string())
    .default([])
    .describe('Ordered list of suggested answers (omit if none).')
    .meta({ usageValue: 'suggested answer here' }),
})

export const toolInfo = {
  name: 'ask_followup_question',
  description:
    'Call this when vital details are missing. Pose each follow-up as one direct, unambiguous question. If it speeds the reply, add up to five short, mutually-exclusive answer options. Group any related questions in the same call to avoid a back-and-forth chain.',
  parameters: z
    .object({
      questions: z
        .array(questionObject)
        .describe('One or more follow-up questions you need answered before you can continue.')
        .meta({ usageValue: 'questions here' }),
    })
    .meta({
      examples: [
        {
          description: 'Single clarifying question (no options)',
          input: {
            questions: { prompt: 'What is the target deployment environment?' },
          },
        },
        {
          description: 'Single question with multiple-choice options',
          input: {
            questions: {
              prompt: 'Which frontend framework are you using?',
              options: ['React', 'Angular', 'Vue', 'Svelte'],
            },
          },
        },
        {
          description: 'Two related questions in one call',
          input: {
            questions: [
              { prompt: 'What type of application are you building?' },
              {
                prompt: 'Preferred programming language?',
                options: ['JavaScript', 'TypeScript', 'Python', 'Java'],
              },
            ],
          },
        },
        {
          description: 'Binary (yes/no) confirmation',
          input: {
            questions: {
              prompt: 'Is it acceptable to refactor existing tests to improve performance?',
              options: ['Yes', 'No'],
            },
          },
        },
      ],
    }),
  permissionLevel: PermissionLevel.None,
} as const satisfies ToolInfoV2

export const handler: ToolHandler<typeof toolInfo, InteractionProvider> = async (provider, args) => {
  if (!provider.askFollowupQuestion) {
    return {
      type: ToolResponseType.Error,
      message: 'Not possible to ask followup question. Abort.',
    }
  }

  const { questions } = toolInfo.parameters.parse(args)
  if (questions.length === 0) {
    return {
      type: ToolResponseType.Invalid,
      message: 'No questions provided',
    }
  }

  const answers = []
  for (const question of questions) {
    const { prompt, options } = question
    const answer = await provider.askFollowupQuestion(prompt, options)
    answers.push(`<ask_followup_question_answer question="${prompt}">
${answer}
</ask_followup_question_answer>`)
  }

  return {
    type: ToolResponseType.Reply,
    message: answers.join('\n'),
  }
}

export const isAvailable = (provider: InteractionProvider): boolean => {
  return !!provider.askFollowupQuestion
}

export default {
  ...toolInfo,
  handler,
  isAvailable,
} satisfies FullToolInfoV2
