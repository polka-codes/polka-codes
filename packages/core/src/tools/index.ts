export * from './provider'
export * from './allTools'
export { default as replaceInFile } from './replaceInFile' // not included in allTools

export * as allTools from './allTools'
