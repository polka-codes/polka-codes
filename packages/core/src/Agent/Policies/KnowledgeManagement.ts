import { dirname, join } from 'node:path'
import { Policies } from '../../config'
import { listFiles, readFile, writeToFile } from '../../tools'
import { getString, getStringArray } from '../../tools/utils'
import type { AgentPolicy } from '../AgentBase'
import type { AssistantMessageContent } from '../parseAssistantMessage'

const prompt = `
====

# Knowledge Extraction & Maintenance

You are equipped with **Knowledge Management** capabilities:

1. **What to capture**
   • Public API of each file (public classes, functions, methods, parameters, return types).
   • High‑level description of each file’s purpose.
   • Invariants and assumptions that must always hold.
   • Project‑ or directory‑specific coding patterns, styles, and architectural conventions.
   • Rules (commenting, testing, documentation, security, etc.).
   • Any other insight that a future contributor would find crucial.

2. **Where to store it**
   • Save knowledge in a YAML file named \`knowledge.ai.yml\`.
   • **Create the file in the repository root if it does not yet exist.**
   • One file per directory.
     – The repository root file records knowledge that applies project‑wide (e.g., service responsibilities, global patterns).
     – Each sub‑directory keeps only the knowledge relevant to that directory or package.
   • Use clear keys such as \`description\`, \`files\`, \`rules\`.

3. **When to update**
   • **Default behaviour:** only create / update knowledge for the files you actively read, create, or modify during the current task.
     – Operate on other files **only if the user explicitly requests it**.
   • **While working**: after reading, analysing, creating, or modifying code, immediately record any new or changed knowledge.
   • **On refactor / deletion**: locate and delete or amend obsolete entries so that knowledge never drifts from the codebase.
   • **Granularity**: update only the affected directory’s \`knowledge.ai.yml\`, except when the change has global impact.

4. **How to format (illustrative)**
\`\`\`yaml
description: "description of the directory"
files:
  - path: "src/utils/math.ts"
    description: "Numeric helpers for currency calculations"
    api:
      functions:
        - name: "add"
          params: [{ name: "a", type: "number" }, { name: "b", type: "number" }]
          returns: "number"
rules:
  - "rules that applies to all files in this directory"
\`\`\`

5. **Source of truth**
   • **Never invent knowledge.** Everything you record must be *directly derived* from existing code, comments, commit messages, or explicit user instructions.
   • If a section has no confirmed content, omit it rather than guessing.

6. **Automatic context**
   When you are asked to read or modify a file, the orchestration layer will supply any existing knowledge for that path automatically. Use it, refine it, and keep it accurate.

Your workflow **must**:
   1. Detect knowledge deltas.
   2. Create \`knowledge.ai.yml\` if missing and write edits to the correct file.
   3. Remove stale facts.
   4. Use provided tools to update the knowledge files.
   5. Record only evidence‑based information; do not hallucinate.

Adhere to these rules rigorously to ensure the codebase and its living documentation stay in sync.
`

export const KnowledgeManagementPolicy = (tools: Parameters<AgentPolicy>[0]) => {
  if (!tools[readFile.name]) {
    return undefined
  }

  // keep a cache of all knowledge files read to we don't read the same file twice
  const readFiles = new Set<string>()

  return {
    name: Policies.KnowledgeManagement,
    prompt: tools[writeToFile.name] ? prompt : undefined,
    async getKnowledgeFilePaths(inputFiles: string[]) {
      const paths = new Set<string>()

      for (const file of inputFiles) {
        // add all directories to the paths set
        let dir = dirname(file)
        paths.add(dir)

        // add all parent directories to the paths set
        while (dir !== '.') {
          paths.add(dir)
          dir = dirname(dir)
        }
      }

      const allFullPaths = []
      for (const path of paths) {
        if (path === '.') {
          continue
        }
        const fullpath = join(path, 'knowledge.ai.yml')
        if (!readFiles.has(fullpath)) {
          allFullPaths.push(fullpath)
          readFiles.add(fullpath)
        }
      }

      return allFullPaths
    },
    async updateResponse(response: AssistantMessageContent[]) {
      const files = new Set<string>()

      for (const content of response) {
        if (content.type === 'tool_use') {
          switch (content.name) {
            case readFile.name: {
              const paths = getStringArray(content.params, 'path')
              for (const path of paths) {
                files.add(path)
              }
              break
            }
            case listFiles.name: {
              const path = getString(content.params, 'path')
              files.add(path)
              break
            }
          }
        }
      }

      const allFullPaths = await this.getKnowledgeFilePaths(Array.from(files))

      if (allFullPaths.length === 0) {
        return response
      }

      return [
        ...response,
        {
          type: 'tool_use',
          name: readFile.name,
          params: {
            path: allFullPaths.join(','),
          },
        } as const,
      ]
    },
  }
}

export type KnowledgeManagementPolicyInstance = NonNullable<ReturnType<typeof KnowledgeManagementPolicy>>
