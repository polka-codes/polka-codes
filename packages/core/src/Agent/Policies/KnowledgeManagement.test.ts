/**
 * Tests for KnowledgeManagement policy
 * Generated by polka.codes
 */

import { describe, expect, test } from 'bun:test'
import { Policies } from '../../config'
import { readFile, writeToFile } from '../../tools'
import { KnowledgeManagementPolicy } from './KnowledgeManagement'

describe('KnowledgeManagementPolicy', () => {
  test('returns undefined when readFile tool is not available', () => {
    const tools = {}
    const policy = KnowledgeManagementPolicy(tools)
    expect(policy).toBeUndefined()
  })

  describe('with readFile tool', () => {
    const tools = {
      [readFile.name]: readFile,
    }

    test('when writeToFile tool is not available', () => {
      const policy = KnowledgeManagementPolicy(tools)
      expect(policy?.name).toBe(Policies.KnowledgeManagement)
      expect(policy?.prompt).toBeUndefined()
    })

    test('includes prompt when writeToFile tool is available', () => {
      const policy = KnowledgeManagementPolicy({
        ...tools,
        [writeToFile.name]: writeToFile,
      })
      expect(policy?.prompt).toBeDefined()
      expect(policy?.prompt).toMatchSnapshot()
    })

    describe('getKnowledgeFilePaths', () => {
      test('returns correct paths for single file', async () => {
        const policy = KnowledgeManagementPolicy(tools)!
        const paths = await policy.getKnowledgeFilePaths(['src/foo/bar.ts'])
        expect(paths).toEqual(['src/foo/knowledge.ai.yml', 'src/knowledge.ai.yml'])
      })

      test('returns correct paths for multiple files', async () => {
        const policy = KnowledgeManagementPolicy(tools)!
        const paths = await policy.getKnowledgeFilePaths(['src/foo/bar.ts', 'src/baz/qux.ts'])
        expect(paths).toEqual(['src/foo/knowledge.ai.yml', 'src/knowledge.ai.yml', 'src/baz/knowledge.ai.yml'])
      })

      test('ignores root directory', async () => {
        const policy = KnowledgeManagementPolicy(tools)!
        const paths = await policy.getKnowledgeFilePaths(['foo.ts'])
        expect(paths).toEqual([])
      })
    })

    describe('updateResponse', () => {
      test('adds readFile for knowledge files when readFile tool is used', async () => {
        const policy = KnowledgeManagementPolicy(tools)!
        const response = await policy.updateResponse([
          {
            type: 'tool_use',
            name: 'read_file',
            params: { path: 'src/foo/bar.ts' },
          },
        ])
        expect(response).toMatchSnapshot()
      })

      test('adds readFile for knowledge files when listFiles tool is used', async () => {
        const policy = KnowledgeManagementPolicy(tools)!
        const response = await policy.updateResponse([
          {
            type: 'tool_use',
            name: 'list_files',
            params: { path: 'src/foo' },
          },
        ])
        expect(response).toMatchSnapshot()
      })

      test('does not add readFile for other tool uses', async () => {
        const policy = KnowledgeManagementPolicy(tools)!
        const response = await policy.updateResponse([
          {
            type: 'tool_use',
            name: 'write_to_file',
            params: { path: 'src/foo/bar.ts', content: 'test' },
          },
        ])
        expect(response).toMatchSnapshot()
      })

      test('combines paths from multiple tool uses', async () => {
        const policy = KnowledgeManagementPolicy(tools)!
        const response = await policy.updateResponse([
          {
            type: 'tool_use',
            name: 'read_file',
            params: { path: 'src/foo/bar.ts' },
          },
          {
            type: 'tool_use',
            name: 'list_files',
            params: { path: 'src/baz' },
          },
        ])
        expect(response).toMatchSnapshot()
      })
    })
  })
})
