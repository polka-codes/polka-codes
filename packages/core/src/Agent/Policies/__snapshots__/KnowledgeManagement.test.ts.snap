// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`KnowledgeManagementPolicy with readFile tool updateResponse adds readFile for knowledge files when readFile tool is used 1`] = `
[
  {
    "name": "read_file",
    "params": {
      "path": "src/foo/bar.ts",
    },
    "type": "tool_use",
  },
  {
    "name": "read_file",
    "params": {
      "path": "src/foo/knowledge.ai.yml,src/knowledge.ai.yml",
    },
    "type": "tool_use",
  },
]
`;

exports[`KnowledgeManagementPolicy with readFile tool updateResponse adds readFile for knowledge files when listFiles tool is used 1`] = `
[
  {
    "name": "list_files",
    "params": {
      "path": "src/foo",
    },
    "type": "tool_use",
  },
  {
    "name": "read_file",
    "params": {
      "path": "src/knowledge.ai.yml",
    },
    "type": "tool_use",
  },
]
`;

exports[`KnowledgeManagementPolicy with readFile tool updateResponse does not add readFile for other tool uses 1`] = `
[
  {
    "name": "write_to_file",
    "params": {
      "content": "test",
      "path": "src/foo/bar.ts",
    },
    "type": "tool_use",
  },
]
`;

exports[`KnowledgeManagementPolicy with readFile tool updateResponse combines paths from multiple tool uses 1`] = `
[
  {
    "name": "read_file",
    "params": {
      "path": "src/foo/bar.ts",
    },
    "type": "tool_use",
  },
  {
    "name": "list_files",
    "params": {
      "path": "src/baz",
    },
    "type": "tool_use",
  },
  {
    "name": "read_file",
    "params": {
      "path": "src/foo/knowledge.ai.yml,src/knowledge.ai.yml",
    },
    "type": "tool_use",
  },
]
`;

exports[`KnowledgeManagementPolicy with readFile tool includes prompt when writeToFile tool is available 1`] = `
"
====

# Knowledge Extraction & Maintenance

You are equipped with **Knowledge Management** capabilities:

1. **What to capture**
   • Public API of each file (public classes, functions, methods, parameters, return types).
   • High‑level description of each file’s purpose.
   • Invariants and assumptions that must always hold.
   • Project‑ or directory‑specific coding patterns, styles, and architectural conventions.
   • Rules (commenting, testing, documentation, security, etc.).
   • Any other insight that a future contributor would find crucial.

2. **Where to store it**
   • Save knowledge in a YAML file named \`knowledge.ai.yml\`.
   • **Create the file in the repository root if it does not yet exist.**
   • One file per directory.
     – The repository root file records knowledge that applies project‑wide (e.g., service responsibilities, global patterns).
     – Each sub‑directory keeps only the knowledge relevant to that directory or package.
   • Use clear keys such as \`description\`, \`files\`, \`rules\`.

3. **When to update**
   • **Default behaviour:** only create / update knowledge for the files you actively read, create, or modify during the current task.
     – Operate on other files **only if the user explicitly requests it**.
   • **While working**: after reading, analysing, creating, or modifying code, immediately record any new or changed knowledge.
   • **On refactor / deletion**: locate and delete or amend obsolete entries so that knowledge never drifts from the codebase.
   • **Granularity**: update only the affected directory’s \`knowledge.ai.yml\`, except when the change has global impact.

4. **How to format (illustrative)**
\`\`\`yaml
description: "description of the directory"
files:
  - path: "src/utils/math.ts"
    description: "Numeric helpers for currency calculations"
    api:
      functions:
        - name: "add"
          params: [{ name: "a", type: "number" }, { name: "b", type: "number" }]
          returns: "number"
rules:
  - "rules that applies to all files in this directory"
\`\`\`

5. **Source of truth**
   • **Never invent knowledge.** Everything you record must be *directly derived* from existing code, comments, commit messages, or explicit user instructions.
   • If a section has no confirmed content, omit it rather than guessing.

6. **Automatic context**
   When you are asked to read or modify a file, the orchestration layer will supply any existing knowledge for that path automatically. Use it, refine it, and keep it accurate.

Your workflow **must**:
   1. Detect knowledge deltas.
   2. Create \`knowledge.ai.yml\` if missing and write edits to the correct file.
   3. Remove stale facts.
   4. Use provided tools to update the knowledge files.
   5. Record only evidence‑based information; do not hallucinate.

Adhere to these rules rigorously to ensure the codebase and its living documentation stay in sync.
"
`;
